### PR TITLE
Revamp of the Anchore Grype parser

### PR DESCRIPTION
**Pull request by StefanFl**
This PR is basically a rewrite of the Anchore Grype parser:

- The reports contain a lot of information that was not stored in the findings so far. Now the findings are much richer than before.
- Internal deduplication was based on CVE only. When you have several components with the same CVE, they had been stored in one finding under one component name. Now the internal deduplication is done by the title, which contains CVE, component name and component version.

### CODE REVIEW

File: dojo/tools/anchore_grype/parser.py [hunk 1]
```diff
@@ -23,44 +22,143 @@ def get_findings(self, file, test):
...
-                    component_version=purl.version,
-                    vuln_id_from_tool=cve,
-                    nb_occurences=1,
-                )
+                            title=finding_title,
+                            description=finding_description,
+                            cve=finding_cve,
+                            cwe=1352,
+                            cvssv3=finding_cvss3,
+                            cvssv3_score=finding_cvss3_score,
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:149]: If you provide the vector, the score is calculated automatically.
```suggestion
```
So better to provide the vector only.

File: dojo/tools/anchore_grype/parser.py [hunk 2]
```diff
@@ -23,44 +22,143 @@ def get_findings(self, file, test):
...
+            vuln_id = vulnerability["id"]
+            vuln_namespace = vulnerability.get('namespace', None)
+            vuln_datasource = vulnerability.get('dataSource', None)
+            vuln_severity = self._convert_severity(vulnerability["severity"])
+            vuln_urls = vulnerability.get('urls', None)
+            vuln_description = vulnerability.get('description', None)
+            vuln_fix_versions = None
+            fix = vulnerability.get('fix', None)
+            if fix:
+                vuln_fix_versions = fix.get('versions', None)
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:35]:
```suggestion
            if 'fix' in vulnerability:
                vuln_fix_versions = vulnerability.get('fix').get('versions')
```
Better, you can do it in one line:
```
vuln_fix_versions = vulnerability.get('fix', []).get('versions')
```
But the previous line is more readable.

File: dojo/tools/anchore_grype/parser.py [hunk 3]
```diff
@@ -23,44 +22,143 @@ def get_findings(self, file, test):
...
-            severity = self._convert_severity(item["vulnerability"]["severity"])
-            purl = PackageURL.from_string(item["artifact"]["purl"])
-            description = ""
-            description += f"\n**CVE:** {cve}"
-            if type(item["matchDetails"]) is dict:
-                description += f'\n**Matcher:** {item["matchDetails"]["matcher"]}'
+            vulnerability = item['vulnerability']
+            vuln_id = vulnerability["id"]
+            vuln_namespace = vulnerability.get('namespace', None)
+            vuln_datasource = vulnerability.get('dataSource', None)
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:28]: The method `get()` return `None` if the key is not in the dictionary.
So the correct form is:
```suggestion
            vuln_namespace = vulnerability.get('namespace')
            vuln_datasource = vulnerability.get('dataSource')
```

File: dojo/tools/anchore_grype/parser.py [hunk 4]
```diff
@@ -23,44 +22,143 @@ def get_findings(self, file, test):
...
+                rel_urls = related_vulnerability.get('urls', None)
+                rel_description = related_vulnerability.get('description', None)
+                rel_cvss = related_vulnerability.get('cvss', None)
+
+            matches = item['matchDetails']
+
+            artifact = item['artifact']
+            artifact_name = artifact.get('name', None)
+            artifact_version = artifact.get('version', None)
+            artifact_purl = artifact.get('purl', None)
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:57]:
```suggestion
```
If you want the parser to not use anymore PURL data, just remove it completely.

File: dojo/tools/anchore_grype/parser.py [hunk 5]
```diff
@@ -70,3 +168,20 @@ def _convert_severity(self, val):
...
+            return vuln_id
+        elif rel_id and rel_id.startswith('CVE'):
+            return rel_id
+        else:
+            return vuln_id
+
+    def get_cvss(self, cvss):
+        if cvss:
+            for cvss_item in cvss:
+                if cvss_item['version'].startswith('3'):
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:183]: Please consider using the package `cvss` to be aligned with other parsers.
https://defectdojo.github.io/django-DefectDojo/contributing/how-to-write-a-parser/#do-not-parse-cvss-by-hand-vector-score-or-severity

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/anchore_grype/parser.py (hunk 4 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py:57]: The `purl` is still used as part of the description.

File: dojo/tools/anchore_grype/parser.py (hunk 5 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py:183]: The `cvss` item is a list of CVSS2 and CVSS3 entries. The parser needs to find the CVSS3 entry. I don't see how the cvss library helps to find this entry. We can test all cvss vectors with the cvss library, but just using the version number as provided in the json seems to be easier.

File: dojo/tools/anchore_grype/parser.py (hunk 1 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py]: done

File: dojo/tools/anchore_grype/parser.py (hunk 2 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py]: done

File: dojo/tools/anchore_grype/parser.py (hunk 3 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py]: done

File: dojo/tools/anchore_grype/parser.py (hunk 5 above)
Comment (damiencarol) [reply dojo/tools/anchore_grype/parser.py:179]: Using `cvss` module provide us a check of the vectors out-of-the-box which means that the parser could be able to see not-well formatted vectors when reading the report and not crash the upper layers (which check this data with a regex in Django model).

File: unittests/tools/test_anchore_grype_parser.py [hunk 6]
```diff
@@ -86,3 +88,165 @@ def test_grype_parser_with_new_matcher_list(self):
...
+- http://www.trustix.org/errata/2004/0050
+- http://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=136304
+- https://exchange.xforce.ibmcloud.com/vulnerabilities/17583
+- https://oval.cisecurity.org/repository/search/definition/oval%3Aorg.mitre.oval%3Adef%3A10497
+- https://lists.apache.org/thread.html/rc713534b10f9daeee2e0990239fa407e2118e4aa9e88a7041177497c@%3Cissues.guacamole.apache.org%3E'''
+        self.assertEqual(references, finding.references)
+        self.assertEqual('libgssapi-krb5-2', finding.component_name)
+        self.assertEqual('1.17-3+deb10u3', finding.component_version)
+        self.assertEqual('CVE-2004-0971', finding.vuln_id_from_tool)
+        self.assertEqual(['dpkg'], finding.tags)
```
Comment (damiencarol) [inline unittests/tools/test_anchore_grype_parser.py:128]: better use:
```suggestion
        self.assertIn('dpkg', finding.tags)
```
More future proof if we add more tags

Comment (StefanFl) [reply unittests/tools/test_anchore_grype_parser.py:128]: The idea is to check exactly what's in the test data. If someone changes the test data without the test case, I think it's ok when it fails.

File: dojo/tools/anchore_grype/parser.py (hunk 5 above)
Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py]: done

File: unittests/tools/test_anchore_grype_parser.py (hunk 6 above)
Comment (damiencarol) [reply unittests/tools/test_anchore_grype_parser.py:128]: ok

File: dojo/tools/anchore_grype/parser.py [hunk 7]
```diff
@@ -70,3 +166,21 @@ def _convert_severity(self, val):
...
+        else:
+            return vuln_id
+
+    def get_cvss(self, cvss):
+        if cvss:
+            for cvss_item in cvss:
+                vector = cvss_item['vector']
+                cvss_objects = cvss_parser.parse_cvss_from_text(vector)
+                if len(cvss_objects) > 0 and type(cvss_objects[0]) == CVSS3:
+                    vector = cvss_item['vector']
```
Comment (damiencarol) [inline dojo/tools/anchore_grype/parser.py:184]: Line 181 and 184 are the same.

Comment (StefanFl) [reply dojo/tools/anchore_grype/parser.py]: Well spotted, I have removed the second line

**Review by damiencarol — APPROVED** [review]

**Review by devGregA — APPROVED** [review]

**Pull request merged**